frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -211,219 +230,290 @@ describe("QueryBuilder", () => {
-    });
-
-    it("should not trigger beforeunload event when query is unedited", async () => {
-      const { mockEventListener } = await setup({
-        card: TEST_NATIVE_CARD,
-      });
-
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-      expect(mockEvent.returnValue).not.toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
-    });
-  });
-
-  describe("beforeunload events in structured queries", () => {
-    afterEach(() => {
-      jest.restoreAllMocks();
-    });
-
-    it("should not trigger beforeunload event when leaving edited question which will turn the question ad-hoc", async () => {
-      const { mockEventListener } = await setup({
-        card: TEST_STRUCTURED_CARD,
-      });
-
-      expect(screen.queryByText("Count")).not.toBeInTheDocument();
-      userEvent.click(await screen.findByText("Summarize"));
-      userEvent.click(await screen.findByText("Done"));
-      expect(await screen.findByText("Count")).toBeInTheDocument();
+      expect(inputArea).toHaveValue("SELECT 1 union SELECT 2");
 
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-      expect(mockEvent.returnValue).not.toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
-    });
-
-    it("should not trigger beforeunload event when user tries to leave an ad-hoc native query", async () => {
-      const { mockEventListener } = await setup({
-        card: TEST_UNSAVED_STRUCTURED_CARD,
-      });
+      userEvent.click(screen.getByTestId("download-button"));
+      userEvent.click(await screen.findByRole("button", { name: ".csv" }));
 
-      expect(screen.queryByText("Count")).not.toBeInTheDocument();
-      userEvent.click(await screen.findByText("Summarize"));
-      userEvent.click(await screen.findByText("Done"));
-      expect(await screen.findByText("Count")).toBeInTheDocument();
-
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-      expect(mockEvent.returnValue).not.toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
-    });
-
-    it("should not trigger beforeunload event when query is unedited", async () => {
-      const { mockEventListener } = await setup({
-        card: TEST_STRUCTURED_CARD,
-      });
-
-      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
-      expect(mockEvent.preventDefault).not.toHaveBeenCalled();
-      expect(mockEvent.returnValue).not.toEqual(BEFORE_UNLOAD_UNSAVED_MESSAGE);
+      expect(
+        mockDownloadEndpoint.called((url, options) => {
+          const { body: urlSearchParams } = options;
+          const query =
+            urlSearchParams instanceof URLSearchParams
+              ? JSON.parse(urlSearchParams.get("query") ?? "{}")
+              : {};
+          return (
+            url === "/api/dataset/csv" && query?.native.query === "SELECT 1"
```

Comment:
This asserts that the request to download endpoint still uses query from the old result `SELECT 1` as opposed to the new query `SELECT 1 union SELECT 2` which has never run.